app/models/concerns/call_listable.rb

```diff
@@ -8,43 +8,57 @@ module CallListable
   # AND they would otherwise be in the call list
   # (e.g. assigned to current line and in user.patients)
 
-  def recently_called_patients(line)
-    patients.where(line: line)
-            .select { |patient| recently_called_by_user? patient }
+  def ordered_patients(line)
+    call_lists.includes(:patient)
+              .in(line: line)
+              .order_by(order_key: :asc)
+              .map(&:patient)
   end
 
   def call_list_patients(line)
-    patients.where(line: line)
-            .reject { |patient| recently_called_by_user? patient }
+    ordered_patients(line).reject { |x| recently_called_by_user? x }
   end
 
-  def add_patient(patient)
-    patients << patient
-    if call_order
-      reorder_call_list(call_order.unshift(patient.id.to_s))
-    else
-      reorder_call_list([patient.id.to_s])
-    end
-    reload
+  def recently_called_patients(line)
+    ordered_patients(line).select { |x| recently_called_by_user? x }
   end
 
-  def remove_patient(patient)
-    patients.delete patient
+  def recently_reached_patients(line)
+    ordered_patients(line).select { |x| recently_reached_by_user? x }
+  end
+
+  def add_patient(patient)
```

Comment:
This is a slightly different approach to what we've had before - call list objects are super ephemeral and basically meant to be thrown out/regenerated on the regular. I'm very amenable to another approach here, but this seems to be reasonably resilient.

Reply:
my only worry with this approach, is without any sort of transaction lock we could theoretically end up in a space where we've deleted all of this users call list entries for this line and then error and lose the entire call list set. if we're going to loop every existing present call anyway, why not just loop and bump the order_key?

Reply:
I thought about that and wound up not doing it, reasoning that it was easier to mass drop and mass recreate than to update in place. But that might be overkill and the collections are always going to be pretty small, so this might be overoptimization. I'll swap it back to use an update strategy.

(my kingdom for mongodb transactions)